File: 1047.remove-all-adjacent-duplicates-in-string.java
/*
 * @lc app=leetcode id=1047 lang=java
 *
 * [1047] Remove All Adjacent Duplicates In String
 */

// @lc code=start
class Solution {
    public String removeDuplicates(String s) {
        char[] chs = s.toCharArray();
        int w = 0;
        for (int r = 0; r != chs.length; ++r) {
            chs[w] = chs[r];
            if (w > 0 && chs[w] == chs[w - 1]) {
                w -= 1;
            } else {
                w += 1;
            }
        }
        return new String(chs, 0, w);
    }
}
// @lc code=end

File: 1047.remove-all-adjacent-duplicates-in-string.cpp
/*
 * @lc app=leetcode id=1047 lang=cpp
 *
 * [1047] Remove All Adjacent Duplicates In String
 */

// @lc code=start

#include <string>
#include <stack>
#include <vector>
#include <sstream>

using namespace std;

class Solution {
public:
    string removeDuplicates(string s) {
        stack<char> st;
        for (auto ch: s) {
            if (!st.empty() && st.top() == ch) {
                st.pop();
            } else {
                st.push(ch);
            }
        }
        vector<char> vec;
        while (!st.empty()) {
            vec.push_back(st.top());
            st.pop();
        }
        reverse(vec.begin(), vec.end());
        ostringstream oss;
        copy(vec.begin(), vec.end(), ostream_iterator<char>(oss));
        return oss.str();
    }
};
// @lc code=end
